netket/utils/group/_group.py

```diff
@@ -263,50 +263,80 @@ def character_table_readable(self) -> Tuple[List[str], Array]:
 
     @struct.property_cached
-    def _irrep_matrices(self) -> PyTree:
-        random = np.random.standard_normal
-        # Generate the eigensystem of a random linear combination of matrices
-        # in the group's regular representation (namely, product_table == g)
-        # The regular representation contains d copies of each d-dimensional irrep
-        # all of which return the same eigenvalues with eigenvectors that
-        # correspond to one another in the bases of the several irrep copies
-        e, vs = np.linalg.eig(random(len(self))[self.product_table])
-        e = np.stack((e.real, e.imag))
-        # we only need one eigenvector per eigenvalue
-        _, idx = np.unique(comparable(e), return_index=True, axis=1)
-        vs = vs[:, idx]
-        # We will nedd the true product table as well
+    def _irrep_matrices(self) -> List[Array]:
+        """
+        Generates irrep matrices using Dixon's algorithm (Math. Comp. 24 (1970), 707).
+
+        We decompose the regular representation of the group into its irreps
+        For real irreps, we make sure the matrices are real
+        """
+
+        # Check Frobenius-Schur indicators of all irreps
         true_product_table = self.product_table[self.inverse]
+        squares = np.diag(true_product_table)
+        frob = np.array(
+            np.rint(
+                np.sum(self.character_table()[:, squares], axis=1).real / len(self)
+            ),
+            dtype=int,
+        )
+
+        def invariant_subspaces(e):
+            # Construct a Hermitian matrix that commutes with all matrices
+            # in the regular rep.
+            # These matrices obey E_{g,h} = e_{gh^{-1}} for some vector e
+            E = e[self.product_table[np.ix_(self.inverse, self.inverse)]]
+            E = E + E.T.conj()
+
+            # Since E commutes with all the ρ, its eigenspaces reduce the rep
+            # With probability 1, there are no accidental degeneracies
+            # except for complex irreps and real symmetric e.
+            e, v = np.linalg.eigh(E)
+
+            # indices that split v into eigenspaces
+            _, starting_idx = np.unique(comparable(e), return_index=True)
+
+            # Calculate sᴴPv for one eigenvector v per eigenspace, a fixed
+            # random vector s and all regular projectors P
+            # These are calculated as linear combinations of sᴴρv for the
+            # regular rep matrices ρ, which is given by the latter two terms
+            vs = v[:, starting_idx]
+            s = np.random.standard_normal(len(self))
+            s = s[self.product_table[np.ix_(self.inverse, self.inverse)]]
+            proj = self.character_table().conj() @ s @ vs
+            starting_idx = list(starting_idx) + [len(self)]
+            return v, starting_idx, proj
+
+        eigen = {}
+        if np.any(frob == 1):
+            # real irreps: start from a real symmetric invariant matrix
+            e = np.random.standard_normal(len(self))
```

Comment:
Is there any particular reason we need randomness in this algorithm? Is the idea to avoid an accidental symmetry in the irreps? And if so, we can still make sure the algorithm gives the same irres every time

Reply:
> Is the idea to avoid an accidental symmetry in the irreps?

Exactly. The eigenspaces of the matrix `E` are all reps, but we want to make sure they're irreps too.

> My fear is that you save a model that uses a particular set of irreps and then you load it with a different set of irreps and it doesn't work

This is basically the reason we chose to keep defining the GCNN kernels in terms of group elements rather than the group FT. We can't guarantee we return the same irrep matrices every time, as they are equivalent up to unitary basis rotations, and the eigenvectors `eigh` returns in degenerate eigenspaces are essentially random.

> This doesn't matter for convolution but for projection it's important

Projections only use the characters, which are invariant under these unitary rotations.

Reply:
> Projections only use the characters, which are invariant under these unitary rotations.

We may want to implement a learned projection into the degenerate subspace for irreps with d > 1. I've been finding the character projections don't work particularly well for higher dimensional representations

Reply:
> We can't guarantee we return the same irrep matrices every time, 

At least this function should take a `key` then, so that when used with the same seed the results also match even though pseudorandom numbers are used.

We got most other parts of NetKet to give reproducible results when used with the same seed, this should probably not be an exception.

Reply:
> > Projections only use the characters, which are invariant under these unitary rotations.
> 
> We may want to implement a learned projection into the degenerate subspace for irreps with d > 1. I've been finding the character projections don't work particularly well for higher dimensional representations

What do you mean? The character projection is *the* valid projection into an irrep subspace...

> > We can't guarantee we return the same irrep matrices every time,
> 
> At least this function should take a `key` then, so that when used with the same seed the results also match even though pseudorandom numbers are used.
> 
> We got most other parts of NetKet to give reproducible results when used with the same seed, this should probably not be an exception.

Fair enough, although that means we switch to `jax.random` here, right?

Reply:
Hm, that would probably be easiest.

You could also make a random seed from the key and use that to initialize a numpy RNG, but that seems to just add more steps to get the same result.

@PhilipVinc Do you agree?

Reply:
Okay, this has also been a bit of an issue for `character_table`, so I switched that to `jax.random` too. Also, some of the logic is now moved to JAX, which might help speed up the linear algebra (and we need the output in JAX code anyway). However, it's now a bit of a mess between JAX and old numpy, and moving it entirely to JAX would be inconvenient, so I wonder if sticking to the latter (apart from random numbers) might be better?

Reply:
Cool. For sake of simplicity I only switched the random numbers to JAX, but we might want to put some computations on the GPU too?

Reply:
In principle I am happy just drawing the random numbers using JAX for consistency with the rest of the code.

Of course I wouldn't mind a potential performance gain from using JAX beyond that. If you want to try and benchmark this, why not. OTOH, it's not crucial for this PR as far as I'm concerned.

Reply:
The bottleneck seems to be doing relatively small linear algebra calculations in a big for loop, and I can't think of any remotely memory-friendly way of vectorising it. I.e. I doubt GPUs would bring a lot of improvement, so I rather won't spend much dev time on sorting out a clean JAX implementation...